File: examples/ai/core-actions-agent/agent.py
"""Core-preset agent (Python twin of agent.mjs).

    python agent.py <input.docx> "<instruction>" [--tracked] [--out <output.docx>]

Demonstrates the `core` LLM-tools preset from the Python SDK via
create_agent_toolkit — one call returning tools, the evaluated system prompt,
and a dispatcher pre-bound to the preset (receipts with verification).

Requires: pip install superdoc-sdk openai  (import name is `superdoc`).
"""

from __future__ import annotations

import json
import os
import sys

from openai import OpenAI
from superdoc import SuperDocClient, create_agent_toolkit

MODEL = os.environ.get("OPENAI_MODEL", "gpt-5.4")
MAX_TURNS = 16


def parse_args(argv: list[str]) -> tuple[str, str, bool, str]:
    tracked = "--tracked" in argv
    out_path = "out.docx"
    if "--out" in argv:
        out_path = argv[argv.index("--out") + 1]
    positional = [
        a for i, a in enumerate(argv)
        if not a.startswith("--") and (i == 0 or argv[i - 1] != "--out")
    ]
    if len(positional) < 2:
        print('Usage: python agent.py <input.docx> "<instruction>" [--tracked] [--out <output.docx>]')
        sys.exit(1)
    return positional[0], positional[1], tracked, out_path


def main() -> None:
    input_path, instruction, tracked, out_path = parse_args(sys.argv[1:])

    # One call — tools, system prompt, and a pre-bound dispatcher that are
    # guaranteed to agree on preset (and excludeActions, if you narrow).
    kit = create_agent_toolkit({"provider": "openai", "preset": "core"})
    tools, system_prompt, dispatch = kit["tools"], kit["system_prompt"], kit["dispatch"]

    if tracked:
        instruction += (
            '\n\nMake every edit as a tracked change (changeMode: "tracked") '
            "so a reviewer can accept or reject it."
        )

    llm = OpenAI()

    with SuperDocClient() as client:
        doc = client.open({"doc": input_path})
        # Everything after open() runs inside try/finally so an API error or a
        # malformed tool call never leaks the session (mirrors agent.mjs).
        try:
            messages = [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": instruction},
            ]

            for _turn in range(MAX_TURNS):
                response = llm.chat.completions.create(model=MODEL, messages=messages, tools=tools)
                message = response.choices[0].message
                messages.append(message)

                if not message.tool_calls:
                    print(f"\n{message.content or '(no final message)'}")
                    break

                for call in message.tool_calls:
                    try:
                        # Malformed tool-call arguments become an error receipt
                        # the model can read and correct, instead of a crash.
                        call_args = json.loads(call.function.arguments)
                        print(f"  -> {call_args.get('action', call.function.name)} ... ", end="", flush=True)
                        receipt = dispatch(doc, call.function.name, call_args)
                        status = receipt.get("status", "ok") if isinstance(receipt, dict) else "ok"
                        verified = receipt.get("verificationPassed") if isinstance(receipt, dict) else None
                        print(status + (" (verification failed)" if verified is False else ""))
                    except Exception as error:
                        # SuperDocError carries a structured .code — keep the
                        # receipt shape identical to agent.mjs.
                        receipt = {
                            "status": "failed",
                            "error": {"code": getattr(error, "code", None), "message": str(error)},
                        }
                        print(f"error: {getattr(error, 'code', None) or error}")

                    messages.append({
                        "role": "tool",
                        "tool_call_id": call.id,
                        # Receipts can contain non-JSON-serializable values.
                        "content": json.dumps(receipt, default=str),
                    })

            doc.save({"out": out_path, "force": True})
            print(f"\nSaved: {out_path}")
        finally:
            doc.close({"discard": True})


if __name__ == "__main__":
    main()

File: examples/ai/core-actions-agent/agent.mjs
/**
 * Core-preset agent: edit a .docx from a natural-language instruction.
 *
 *   node agent.mjs <input.docx> "<instruction>" [--tracked] [--out <output.docx>]
 *
 * Demonstrates the `core` LLM-tools preset end to end via createAgentToolkit:
 *   - tools        → 2 tools (superdoc_inspect, superdoc_perform_action)
 *   - systemPrompt → the prompt the action surface is evaluated with
 *   - dispatch     → pre-bound to the preset; returns receipts with verification
 *
 * Each tool call is printed as a status line — the same shape you would
 * stream to a chat UI (see the "Streaming status to your UI" docs section).
 */
import 'dotenv/config';
import OpenAI from 'openai';
import { createSuperDocClient, createAgentToolkit } from '@superdoc-dev/sdk';

const MODEL = process.env.OPENAI_MODEL ?? 'gpt-5.4';
const MAX_TURNS = 16;

// ---------------------------------------------------------------------------
// CLI arguments
// ---------------------------------------------------------------------------
const args = process.argv.slice(2);
const tracked = args.includes('--tracked');
const outFlag = args.indexOf('--out');
if (outFlag >= 0 && (args[outFlag + 1] == null || args[outFlag + 1].startsWith('--'))) {
  console.error('--out requires a path argument');
  process.exit(1);
}
const outPath = outFlag >= 0 ? args[outFlag + 1] : 'out.docx';
const positional = args.filter((a, i) => !a.startsWith('--') && args[i - 1] !== '--out');
const [inputPath, instruction] = positional;

if (!inputPath || !instruction) {
  console.error('Usage: node agent.mjs <input.docx> "<instruction>" [--tracked] [--out <output.docx>]');
  process.exit(1);
}

// ---------------------------------------------------------------------------
// SuperDoc session + core-preset tool surface
// ---------------------------------------------------------------------------
const client = createSuperDocClient();
await client.connect();
const doc = await client.open({ doc: inputPath });

// Everything after open() runs inside try/finally so an API error or a
// malformed response never leaks the session or (in dev checkouts) the
// locally spawned CLI process.
try {

  // One call — tools, system prompt, and a pre-bound dispatcher that are
  // guaranteed to agree on preset (and excludeActions, if you narrow).
  const { tools, systemPrompt, dispatch } = await createAgentToolkit({ provider: 'openai', preset: 'core' });

  const userInstruction = tracked
    ? `${instruction}\n\nMake every edit as a tracked change (changeMode: "tracked") so a reviewer can accept or reject it.`
    : instruction;

  const messages = [
    { role: 'system', content: systemPrompt },
    { role: 'user', content: userInstruction },
  ];

  // ---------------------------------------------------------------------------
  // Agent loop
  // ---------------------------------------------------------------------------
  const openai = new OpenAI();

  for (let turn = 0; turn < MAX_TURNS; turn += 1) {
    const response = await openai.chat.completions.create({ model: MODEL, messages, tools });
    const message = response.choices[0].message;
    messages.push(message);

    if (!message.tool_calls?.length) {
      console.log(`\n${message.content ?? '(no final message)'}`);
      break;
    }

    for (const call of message.tool_calls) {
      let receipt;
      try {
        // Malformed tool-call arguments become an error receipt the model can
        // read and correct, instead of crashing the run.
        const callArgs = JSON.parse(call.function.arguments);
        process.stdout.write(`  → ${callArgs.action ?? call.function.name} … `);
        receipt = await dispatch(doc, call.function.name, callArgs);
        const status = receipt?.status ?? 'ok';
        const verified = receipt?.verificationPassed;
        console.log(status + (verified === false ? ' (verification failed)' : ''));
      } catch (error) {
        receipt = { status: 'failed', error: { code: error.code, message: error.message } };
        console.log(`error: ${error.code ?? error.message}`);
      }

      messages.push({
        role: 'tool',
        tool_call_id: call.id,
        content: JSON.stringify(receipt),
      });
    }
  }

  // ---------------------------------------------------------------------------
  // Save
  // ---------------------------------------------------------------------------
  await doc.save({ out: outPath, force: true });
  console.log(`\nSaved: ${outPath}`);

} finally {
  await doc.close({ discard: true }).catch(() => {});
  await client.dispose();
}
